src/embeddedContentFileSystemProvider.ts

```diff
@@ -0,0 +1,332 @@
+            if (binaryContentRenderer) {
+                return Buffer.from(binaryContentRenderer.renderAsMarkdown(parsedUriData.originalFileName), 'utf8');
+            }
+        }
+
+        if (artifact.contents.rendered) {
+            if (artifact.contents.rendered.markdown) {
+                return Buffer.from(artifact.contents.rendered.markdown, artifact.encoding);
+            }
+
+            if (artifact.contents.rendered.text) {
+                return Buffer.from(artifact.contents.rendered.text, artifact.encoding);
+            }
+        }
+
+        throw new Error(`There is no contents that can be rendered associated with artifact index ${parsedUriData.runIndex} for run index ${parsedUriData.runIndex}.`);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public writeFile(uri: vscode.Uri, content: Uint8Array, options: { create: boolean; overwrite: boolean }): void {
+        throw new Error('Not implemented');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public delete(uri: vscode.Uri, options: { recursive: boolean }): void {
+        throw new Error('Not implemented');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public rename(oldUri: vscode.Uri, newUri: vscode.Uri, options: { overwrite: boolean }): void {
+        throw new Error('Not implemented');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public dispose(): void {
+        vscode.Disposable.from(...this.disposables).dispose();
+        this.disposables = [];
+    }
+
+    /**
+     * Creates an embedded content URI.
+     * @param sarifLog The raw sarif log.
+     * @param logPath The full path to the SARIF log file.
+     * @param fileName The file name that VSCode will display in the editor and use for detection of type.
+     * @param runIndex The index of the run in the SARIF file.
+     * @param artifactIndex The artifact index.
+     */
+    public static tryCreateUri(sarifLog: sarif.Log, logPath: vscode.Uri, fileName: string, runIndex: number, artifactIndex: number): vscode.Uri | undefined {
+        if (!logPath.isSarifFile()) {
+            throw new Error(`${logPath.toString()} is not a SARIF file`);
+        }
+
+        const run: sarif.Run | undefined = sarifLog.runs[runIndex];
+        if (!run || !run.artifacts) {
+            return undefined;
+        }
+        const artifact: sarif.Artifact | undefined = run.artifacts[artifactIndex];
+        if (!artifact || !artifact.contents) {
+            return undefined;
+        }
+
+        // We render binary and markdown content as markdown, so
+        // we add the ".md" extension for VSCode to detect that.
+        const uriFileName: string = (artifact.contents.binary  || artifact.contents.rendered?.markdown) ? `${fileName}.md` : fileName;
+        const originalFileNameAsBase64: string = new Buffer(fileName).toString('base64');
+        const logPathAsBase64: string = new Buffer(logPath.toString(/*skipEncoding*/ true), 'UTF8').toString('base64');
```

Comment:
Why Base64 encoding paths versus URI encoding (`encodeURI()`)? The Base64 encoding can still be troublesome during Url.parse().